src/persist/src/s3.rs

```diff
@@ -397,87 +397,63 @@ impl Blob for S3Blob {
-            min_body_elapsed_.observe(elapsed, "s3 download first part body");
-
-            result
-        };
-
-        let mut body_handles = Vec::new();
-        // TODO: Add the key and part number once this can be annotated with metadata.
-        body_handles.push(async_runtime.spawn_named(|| "persist_s3blob_get_body", first_body_fut));
-
-        if num_parts > 1 {
-            // Fetch the headers of the rest of the parts. (Starting at part 2 because we already
-            // did part 1.)
-            for part_num in 2..=num_parts {
-                let header_future = self
-                    .client
-                    .get_object()
-                    .bucket(&self.bucket)
-                    .key(&path)
-                    .part_number(part_num)
-                    .send();
-
-                // Clone a handle to our MinElapsed trackers so we can give one to
-                // each download task.
-                let min_header_elapsed_ = Arc::clone(&min_header_elapsed);
-                let min_body_elapsed_ = Arc::clone(&min_body_elapsed);
-
-                let request_future = async move {
-                    // Request our headers.
-                    let header_start = Instant::now();
-                    let object = header_future
-                        .await
-                        .map_err(|err| ExternalError::from(anyhow!("s3 get meta err: {}", err)))?;
-                    let header_elapsed = header_start.elapsed();
-                    min_header_elapsed_.observe(header_elapsed, "s3 download part header");
-
-                    // Request the body.
-                    let body_start = Instant::now();
-                    let body = object
-                        .body
-                        .collect()
-                        .await
-                        .map_err(|err| Error::from(format!("s3 get body err: {}", err)))?;
-                    let body_elapsed = body_start.elapsed();
-                    min_body_elapsed_.observe(body_elapsed, "s3 download part body");
-
-                    Ok(body)
+        let mut body_futures = FuturesOrdered::new();
+        let mut first_part = Some(first_part);
+
+        // Fetch the headers of the rest of the parts. (Starting at part 2 because we already
+        // did part 1.)
+        for part_num in 1..=num_parts {
+            // Clone a handle to our MinElapsed trackers so we can give one to
+            // each download task.
+            let min_header_elapsed = Arc::clone(&min_header_elapsed);
+            let min_body_elapsed = Arc::clone(&min_body_elapsed);
+            let first_part = first_part.take();
+            let path = &path;
+            let request_future = async move {
+                // Fetch the headers of the rest of the parts. (Using the existing headers
+                // for part 1.
+                let object = match first_part {
+                    Some(first_part) => first_part,
```

Comment:
this branch is taken if and only if part_num == 1, correct? if so, let's stick a `debug_assert_eq!(part_num, 1)` in here with a comment about that invariant. (took me a sec to wrap my head around how this worked when reviewing)